Add ObjectLiteralConfiguration with color/image options

Dschee opened the pull request:
This fixes #1587 by implementing the discussed option configuration.

Dschee commented:
No idea what to do to prevent Circle CI from failing – Danger tells me I need to configure `DANGER_GITHUB_API_TOKEN` but I can't find any info about that in the CONTRIBUTING.md, so not sure what to do now. Where to I add that? Do I really need it? (Other PRs seem all to run fine.) And what value should it have? I'm confused ... ^^

marcelofabri commented:
@Dschee that is not your fault, but shouldn't happen anyway. @jpsim, do you know why would it fail?

Note that the Travis failure is a valid one tough.

Dschee commented:
Uh, yeah, the Travis failure was due to the date, which was formed in the German style by my Xcode automatically (`03.06.17`). Just fixed it and used the weird American style instead (`06/03/17` jumping from month to day and back to year ^o^). Should pass now, hopefully!

SDGGiesbrecht commented:
Merhaba, @Dschee, katkınız için teşekkür ederim. (Ich kann nur wenig Türkisch – hoffentlich bedeutet das tatsächlich was ich sagen wollte...)

> [It] was due to the date [...], which was formed in German style. [...] Just fixed it and used the weird American style instead[.]

Nach mir wären es auch eigentlich die anderen Daten die nicht stimmen. 😄 Dein lustiger Kommentar hat mich gefreut.

Dschee commented:
@SDGGiesbrecht Haha, ja, ich denke das passt schon so, wobei statt "katkı" (eher für Rezepte verwendet) besser "yardım" oder "ekleme" genommen werden sollte. Aber hey, sämtliche Konjugation ist richtig. ;)

Bzgl. des Datums muss ich zugeben, dass wir im Deutschen auch solche unsinnigen Dinge haben, Beispiel die Lesart von dreistelligen Zahlen wie 365 – "drei hundert", dann "fünf und", dann Sprung zurück zu "sechzig". Also ist das Deutsche auch nicht unbedingt besser ... 😅

Dschee commented:
@marcelofabri Anything here that still need to be done? I fixed the one valid issue already and consider this merge-ready.

Dschee commented:
@marcelofabri The requested changes were made:

- I've added a new ObjectLiteralRuleTests file (there was none before)
- I've implemented tests for three cases: only color_literal, only image_literal, both together
- Using TDD I've added usage of the options within the rule (sorry for not seeing this before)

I've run the new tests both before it was implemented (with the only color_literal / only image_literal tests failing) and after it was implemented where the tests succeeded.

I hope this is now ready! 🎉

marcelofabri commented:
> I've added a new ObjectLiteralRuleTests file (there was none before)

There was a test in `RulesTests` that now need to be removed. You should also install https://github.com/krzysztofzablocki/Sourcery and run `make sourcery` to make sure that the Linux tests are updated. More info on https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#code-generation

Dschee commented:
I've removed the code from the `RulesTests` but I actually couldn't install sourcery, the installation instruction are really poor or the method is, I don't know. It didn't work and for now I don't have the time to deal with sourcery.

Also `script/cibuild` failed with `make: *** [test] Error 65` on `	PhaseScriptExecution Embed\ Swift\ libraries\ into\ SwiftLintFramework.framework` which doesn't give a lot of information to look into, really.

Next, when trying `swift test` I got `Can't parse Package.swift manifest file because it contains invalid format. Fix Package.swift file format and try again.`. No idea, I've not even opened Package.swift...

And `make docker_test` obviously fails because I don't have it installed. 

I'm not sure if I understood "When contributing code changes, please ensure that all three supported build methods continue to work and pass tests." correctly – cause if I really had to get all of them work and understand what they all do, that would be quite a big amount of work for contributing to this project. :D

marcelofabri commented:
> I've removed the code from the RulesTests but I actually couldn't install sourcery, the installation instruction are really poor or the method is, I don't know. It didn't work and for now I don't have the time to deal with sourcery.

`brew install sourcery` should install it.

> Also script/cibuild failed with make: *** [test] Error 65 on PhaseScriptExecution Embed\ Swift\ libraries\ into\ SwiftLintFramework.framework which doesn't give a lot of information to look into, really.
> Next, when trying swift test I got Can't parse Package.swift manifest file because it contains invalid format. Fix Package.swift file format and try again.. No idea, I've not even opened Package.swift...

Are you using Xcode 8.3? The error message is probably from Xcode itself, so ¯\_(ツ)_/¯ 

> And make docker_test obviously fails because I don't have it installed.

You can install it by downloading from the website.

> I'm not sure if I understood "When contributing code changes, please ensure that all three supported build methods continue to work and pass tests." correctly – cause if I really had to get all of them work and understand what they all do, that would be quite a big amount of work for contributing to this project. Is there no simpler way?

Testing everything locally is the perfect scenario, since it avoids surprises on CI and also helps you to understand better the issue and debug it if needed. However, these checks all run on CI anyway. So unless there's something wrong, usually just running from Xcode is enough.

Dschee commented:
Thanks for the help, they should really document that it's installable via Homebrew. But now that I run `make sourcery` I just get `make: `sourcery' is up to date.`. I got it so far earlier, too. But I thought that it was an issue, since I expected to see some warnings/errors or code changes happen. Nothing like that though.

Yeah, you got me, I had Xcode 9.0 beta configured on the command line. :-D Sorry for that, switching back and forth between the two recently to try out new stuff. ^^

I'm gonna wait for the CI to check the docker part, I doubt that my changes have an influence on that, but we'll see ... I can install it later if I need to. Thanks for explaining all to me, makes all sense with the CI in there. This is just not documented like this in the CONTRIBUTING.md. ;)

Dschee commented:
Ok, docker actually failed, probably due to the `make sourcery` issue, I guess?

marcelofabri commented:
> I'm gonna wait for the CI to check the docker part, I doubt that my changes have an influence on that

Actually, that's the main issue: you're changes are not being tested on Linux because [LinuxMain.swift](https://github.com/realm/SwiftLint/blob/master/Tests/LinuxMain.swift) wasn't updated.

> Ok, docker actually failed, probably due to the make sourcery issue, I guess?

Yes, because the list of tests from `RulesTests` wasn't updated after you removed that test.

Dschee commented:
Yess, finally I got it to work! I still have no idea why `make sourcery` doesn't work for me. But running only `make` also ran the sourcery step and everything updated just fine. Now it should pass, I hope!

marcelofabri commented on CHANGELOG.md:
```diff
@@ -39,6 +39,12 @@
 
+* Add `code_literal` and `image_literal` boolean configuration options to
+  `object_literal` rule. They allow to only check for one or the other
+  literal type instead of both together.
```
requires two trailing spaces as described in CONTRIBUTING.md: https://github.com/realm/SwiftLint/blob/master/CONTRIBUTING.md#tracking-changes

Dschee replied:
Fixed.

marcelofabri commented on Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift:
```diff
@@ -0,0 +1,94 @@
+        verifyRule(ObjectLiteralRule.description)
+    }
+
+    // swiftlint:disable:next function_body_length
```
I don't think you need any of these `disable:next`

Dschee replied:
Fixed.

marcelofabri commented on Tests/SwiftLintFrameworkTests/ObjectLiteralRuleTests.swift:
```diff
@@ -0,0 +1,94 @@
+            identifier: baseDescription.identifier,
+            name: baseDescription.name,
+            description: baseDescription.description,
+            nonTriggeringExamples: baseDescription.nonTriggeringExamples + nonTriggeringColorLiteralExamples,
```
`baseDescription.nonTriggeringExamples` already contains examples similar to `nonTriggeringColorLiteralExamples`

Dschee commented:
@marcelofabri Fixed both of your review comments.

marcelofabri commented:
Thanks! 💯
